### DISCUSSION TITLE
Update atom object value using a custom hook causes multiple re-renders

### DISCUSSION THREAD

**Discussion by dani-z** [pmndrs/jotai#1541] [Q&A]
Hi all,

I am using Jotai inside a React Native app and want to move some logic from a Provider to an Atom + a custom hook. I am unsure if I am doing something wrong, but I see 3 re-renders of the component when I toggle it. Here's a snippet of the code I am using.

Has anyone seen this re-render issue in similar cases before?

Many thanks 🙏 

```js
// atom.tx

export type Biometrics = {
  isBiometricsSupported: boolean;
  biometricsType: Keychain.BIOMETRY_TYPE | null;
  isBiometricsLoading: boolean;
};

const defaultState: Biometrics = {
  isBiometricsSupported: false,
  isBiometricsLoading: true,
  biometricsType: null,
};

export const biometricsAtom = atom<Biometrics>(defaultState);

export const biometricsEnabledAtom = atomWithMMKV<boolean>('is-biometrics-enabled', false);
```

```js
// useBiometrics.ts
export const useBiometrics = () => {
  const [biometrics, setBiometrics] = useAtom(biometricsAtom);
  const [isEnabled, setIsEnabled] = useAtom(biometricsEnabledAtom);
  const passcode = useAtomValue(passcodeAtom);

  const init = async () => {
    const type = await Keychain.getSupportedBiometryType();

    if (!type) {
      setBiometrics((prev) => ({ ...prev, isBiometricsLoading: false }));
      return;
    }

    setBiometrics((prev) => ({
      ...prev,
      isBiometricsSupported: true,
      isBiometricsLoading: false,
      biometricsType: type,
    }));
  };

  const enable = async () => {
    setBiometrics((prev) => ({ ...prev, isBiometricsLoading: true }));
    try {
      await Keychain.setGenericPassword(config.biometricUsername, passcode, {
        service: config.biometricKey,
        accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_ANY,
        accessible: Keychain.ACCESSIBLE.WHEN_PASSCODE_SET_THIS_DEVICE_ONLY,
      });

      setBiometrics((prev) => ({ ...prev, isBiometricsLoading: false }));
      setIsEnabled(true);
    } catch (error) {
      setBiometrics((prev) => ({ ...prev, isBiometricsLoading: false }));
      throw error;
    }
  };

  const disable = async () => {
    setBiometrics((prev) => ({ ...prev, isBiometricsLoading: true }));
    try {
      await Keychain.resetGenericPassword({ service: config.biometricKey });

      setBiometrics((prev) => ({ ...prev, isBiometricsLoading: false }));
      setIsEnabled(false);
    } catch (error) {
      setBiometrics((prev) => ({ ...prev, isBiometricsLoading: false }));
      throw error;
    }
  };

  const toggle = async () => (isEnabled ? disable() : enable());

  const prompt = async () => {
    try {
      return Keychain.getGenericPassword({
        service: config.biometricKey,
      });
    } catch (error) {
      throw error;
    }
  };

  return { init, enable, disable, prompt, isEnabled, toggle, ...biometrics };
};
```

```js
// Component

const BiometricsToggle = memo(() => {
  const {
    toggle,
    isBiometricsSupported,
    isEnabled: isBiometricsEnabled,
  } = useBiometrics();

  return (
    <Switch
        onTrackColor="primary.500"
        disabled={!isBiometricsSupported}
        value={isBiometricsEnabled}
        onToggle={toggle}
      />
  );
});
```

**Comment by dai-shi** [comment]
If it's renders without commits, it's intentional behavior. You can confirm with `useEffect(() => console.log(...))`.

**Reply by dani-z** [reply]
Hi @dai-shi thanks for your reply. You mean to try and add the `useEffect` in the Switch component?

**Reply by dai-shi** [reply]
I mean, how do you determine "multiple-rerenders"? Do you use React DevTools to highlight renders?
What's confusing is "rendering" means two thing in React. Render phase is when React run component functions. Commit phase is when React apply changes to DOM.

**Reply by dani-z** [reply]
Ah, right yes. So I realised something is off when the Switch component was behaving "weird" and just placed a console log inside it (and also inside the parent component). The log was displayed 3 times for each action.

**Reply by dai-shi** [reply]
If you put the console log in useEffect, and the result is expected, there's no issues as it's intentional behavior. (We will be revisiting this intentional behavior in the new api #1515.)

Curious how it was "weird" though.

**Reply by dani-z** [reply]
Yeah it's hard to explain "weird" :D That's just a UI glitch, like the switch doesn't feel natural. When I just used a normal true | false `useState` it worked smoothly. That was the first evidence of something being off. I've placed the console logs outside `useEffect`, which is why I initially referred to it as "re-render".

**Reply by dai-shi** [reply]
Interesting. If you can reproduce the UI glitch in codesandbox https://react.new, I would love to see it.
